daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala

```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
+
+                      case NodeTypeCase.FETCH =>
+                        // A fetch may cause a divulgence, which is why the target contract is a potential output.
+                        ValueCoder.CidDecoder.decode(node.getFetch.getContractIdStruct).map {
+                          contractId => contractIdsOrKeys + ContractIdOrKey.Id(contractId)
+                        }
+
+                      case NodeTypeCase.LOOKUP_BY_KEY =>
+                        // Contract state only modified on divulgence, in which case we'll have a fetch node,
+                        // so no outputs from lookup node.
+                        Right(contractIdsOrKeys)
+
+                      case NodeTypeCase.NODETYPE_NOT_SET =>
+                        Left(ValueCoder.DecodeError("NODETYPE_NOT_SET not supported"))
+                    }
+                  }
+                case (Left(error), _) => Left(error)
+              }
+          }
+          .left
+          .map(ConversionError.DecodeError)
+    }
+
+  def filterCreateAndExerciseNodes(
+      rawTransaction: RawTransaction
+  ): Either[ConversionError, RawTransaction] =
+    Try(TransactionOuterClass.Transaction.parseFrom(rawTransaction.byteString)) match {
+      case Failure(throwable) => Left(ConversionError.ParseError(throwable.getMessage))
+      case Success(transaction) =>
+        val nodes = transaction.getNodesList.asScala
+        val nodeMap: Map[String, TransactionOuterClass.Node] =
+          nodes.view.map(n => n.getNodeId -> n).toMap
+
+        @tailrec
+        def goNodesToKeep(
+            todo: List[String],
+            result: Set[String],
+        ): Either[ConversionError, Set[String]] = todo match {
+          case Nil => Right(result)
+          case head :: tail =>
+            nodeMap.get(head) match {
+              case Some(node) =>
+                node.getNodeTypeCase match {
+                  case NodeTypeCase.CREATE =>
+                    goNodesToKeep(tail, result + head)
+                  case NodeTypeCase.EXERCISE =>
+                    goNodesToKeep(
+                      node.getExercise.getChildrenList.asScala.toList ++ tail,
+                      result + head,
+                    )
+                  case NodeTypeCase.ROLLBACK | NodeTypeCase.FETCH | NodeTypeCase.LOOKUP_BY_KEY |
+                      NodeTypeCase.NODETYPE_NOT_SET =>
+                    goNodesToKeep(tail, result)
+                }
+              case None => Left(ConversionError.InternalError(s"Invalid transaction node id $head"))
+            }
+        }
+
+        goNodesToKeep(transaction.getRootsList.asScala.toList, Set.empty).map { nodesToKeep =>
+          val filteredRoots = transaction.getRootsList.asScala.filter(nodesToKeep)
+
+          def stripUnnecessaryNodes(node: TransactionOuterClass.Node): TransactionOuterClass.Node =
```

Comment:
maybe move that somewhere out. Having a definition nested this deeply is a bit hard to follow.

Reply:
I just inlined it, as it's trivial and used just once.